chai/cmd_executor.py

```diff
@@ -0,0 +1,225 @@
+    """Records a parsing error
+
+    Attributes:
+        errors: A list of parsing error messages.
+    """
+
+    errors: List[str]
+
+    @classmethod
+    def _of_dict(cls, d: dict) -> ParsingError:
+        _ensure_is_pass_failure(d)
+
+        data = d["data"]
+        pass_name = data["pass_name"]
+        if pass_name == "SanyParser":
+            return ParsingError(pass_name, data)
+        else:
+            raise UnexpectedErrorException(
+                f"Unexpected error receieved from RPC call: {d}"
+            )
+
+
+@dataclass
+class TypecheckingError(CmdExecutorError):
+    """Records a typechecking error
+
+    Attributes:
+        errors: A list of tuples pairing source locations with type error
+            messages.
+    """
+
+    errors: List[Tuple[str, str]]  # location, msg errors
+
+    @classmethod
+    def _of_dict(cls, d: dict) -> ParsingError | TypecheckingError:
+        _ensure_is_pass_failure(d)
+
+        pass_name = d["data"]["pass_name"]
+        errors = d["data"]["error_data"]
+        if pass_name == "TypeCheckerSnowcat":
+            return TypecheckingError(pass_name, errors)
+        else:
+            return ParsingError._of_dict(d)
+
+
+@dataclass
+class CheckingError(CmdExecutorError):
+    """Records a model checking error
+
+    Attributes:
+        checking_result: The kind of model checking result. The possible result
+            kinds are as follows
+
+            - Error: A checking violation is found.
+            - Deadlock: A deadlock was found.
+            - RuntimeError: A runtime error was encountered, preventing checking.
+        counter_examples: A list of counterexamples found.
+    """
+
+    checking_result: str
+    counter_example: List[Counterexample]
+
+    @classmethod
+    def _of_dict(cls, d: dict) -> ParsingError | TypecheckingError | CheckingError:
+        _ensure_is_pass_failure(d)
+
+        error_data = d["data"]["error_data"]
+        pass_name = d["data"]["pass_name"]
+        if pass_name == "BoundedChecker":
+            checking_result = error_data["checking_result"]
+            if checking_result == "Deadlock":
+                # TODO We should use the same key for both counterexamples
+                counter_examples = error_data["counterexample"]
+            else:
+                counter_examples = error_data["counterexamples"]
+            # TODO Handle all other checking errors
+            return CheckingError(pass_name, checking_result, counter_examples)
+        else:
+            return TypecheckingError._of_dict(d)
+
+
+E = TypeVar("E")
+CmdExecutorResult = RpcResult[E | TlaModule]
+
+
+def _str_source(spec, aux):
+    return {"source": {"type": "string", "content": spec, "aux": aux}}
+
+
+def _config_json(spec: Input, aux: Optional[List[Input]], cfg: Optional[dict]) -> str:
+    # TODO: error if source already set in config?
+    aux = aux or []
+    config = cfg or {}
+    return json.dumps(config | {"input": _str_source(spec, aux)})
```

Comment:
Seems to be used only once, shall we inline it below?

```suggestion
def _config_json(spec: Input, aux: Optional[List[Input]], cfg: Optional[dict]) -> str:
    # TODO: error if source already set in config?
    aux = aux or []
    config = cfg or {}
    str_source = {"source": {"type": "string", "content": spec, "aux": aux}}
    return json.dumps(config | {"input": str_source})
```